cql3/Cql.g

```diff
@@ -1621,6 +1621,7 @@ property[cql3::statements::property_definitions& props]
 propertyValue returns [sstring str]
     : c=constant           { $str = c->get_raw_text(); }
     | u=unreserved_keyword { $str = u; }
+    | K_NULL { $str = "null"; }
```

Comment:
What's the difference between x = NULL and x = 'null'? Won't they return the same thing?

Maybe propertyValue should return std::optional<sstring> to indicate null.

Reply:
Yeah, but that change would be rather intrusive... It isn't technically different from a string saying "null" right now, and my intention here was just syntactic sugar - being able to use `timeout = null` rather than weird `timeout = "null"`. There's also no danger for this null being ambiguous (at least in terms of timeouts), because `"null"` is not a valid timeout in any case.

Reply:
... and to back up my dubious claim that it's not a big issue (for the record - I also think it's rather dubious), unreserved keywords are treated pretty much the same way - so if somebody really wants to set `timeout = writetime`, which makes no sense, it would work just as good as `timeout = "writetime"`. My intention here was to treat null as unreserved in the particular case of property values.

Reply:
Ugh, the unreserved_keyword thing smells.

I worry that we'll sometimes fix it, then break unintentional code that quoted "null". Maybe use a string that's harder to get accidentally? e.g. "!?null?!" or something.

Reply:
Well, in case of timeouts, I can drop the patch that treats "null" as unreserved if we agree on another way for resetting a timeout value to default. Does `timeout = ""` sound acceptable? It doesn't sound too bad, although null sounded a little more standard.

Reply:
No, NULL is the right thing here.

Reply:
Also, please add a FIXME since the grammar should return a more complicated type (probably an expression that can be evaluated at compile type, or maybe just optional<string>.

Reply:
There's also a less intrusive way I came up with just now - instead of reusing propertyValue from the grammar, I can wrap it in a nullablePropertyValue, based on optionals, and use it just in the service level code.

Reply:
Yes

Reply:
... except is going to be really ugly, I just tried and gave up quickly. The whole `property_definitions` class is based on a map of strings, so it would either need to be templatized, or unconditionally use optionals underneath, or copy-pasted, and all of these options are a little too large to include them in this series for my taste - it's already a rather big change.

So instead, I suggest the following:
1. Add a FIXME in this series
2. Create an issue explaining why strings are not good enough to represent property values (they kind of are now, but they're not future-proof, because `null` is indistinguishable from `"null"` and same goes for all unreserved keywords).
3. Fix the above in a follow-up patch, outside of scope of this one. The CQL layer will remain identical from the user's point of view, but the underneath implementation would become more future-proof once (2) is fixed

Reply:
Ok.

Reply:
https://github.com/scylladb/scylla/issues/8598

Reply:
@avikivity @psarna what about `with timeout = infinity`? If I read the parser correctly, it accepts the keyword `infinity` as an infinite (floating point) constant so it sounds a reasonable choice, and it doesn't require this `unreserved_keyword` thing (maybe we'll even be lucky and can remove this `unreserved_keyword` production from `property`?).

Reply:
Using infinity sounds like a nice potential extension, but for a different purpose. `null` is about deleting the per-service-level timeout and defaulting back to the one defined in `scylla.yaml`. Setting it to infinity sounds more like actually setting the operation's timeout to "never time this operation out".

Reply:
(and the latter can currently be worked around quite neatly by setting the timeout to an unbelievably high value)